frontend/components/SettingsPage/RecoveryModal.tsx

```diff
@@ -0,0 +1,240 @@
+  }).then((res) => {
+    if (res.ok) return res.json();
+    throw new Error('Failed to get recovery seed phrase');
+  });
+
+type ModalStep = 'password' | 'recoveryPhrase';
+
+export const RecoveryModal = ({ open, onClose }: RecoveryModalProps) => {
+  const [step, setStep] = useState<ModalStep>('password');
+  const [password, setPassword] = useState('');
+  const [recoveryPhrase, setRecoveryPhrase] = useState<string[]>([]);
+  const [isCopied, setIsCopied] = useState(false);
+  const [isValid, setIsValid] = useState(false);
+  const { isLoading, validatePassword } = useValidatePassword();
+  const { markAsBackedUp } = useRecoveryPhraseBackup();
+
+  // Reset state when modal closes
+  useEffect(() => {
+    if (!open) {
+      setStep('password');
+      setPassword('');
+      setRecoveryPhrase([]);
+      setIsCopied(false);
+      setIsValid(false);
+    }
+  }, [open]);
+
+  useEffect(() => {
+    if (!password || step !== 'password') {
+      setIsValid(false);
+      return;
+    }
+
+    const validate = async () => {
+      const result = await validatePassword(password);
+      setIsValid(result);
+    };
+
+    const timeoutId = setTimeout(validate, 500);
+    return () => clearTimeout(timeoutId);
+  }, [password, step, validatePassword]);
+
+  const handleReveal = async () => {
+    const result = await validatePassword(password);
+    if (!result) return;
+
+    try {
+      const data = await getRecoverySeedPhrase(password);
+      const mnemonic = data.mnemonic || '';
+      if (mnemonic) {
+        setRecoveryPhrase(mnemonic);
+        setStep('recoveryPhrase');
+      } else {
+        throw new Error('Recovery phrase not found in response');
+      }
+    } catch (e) {
+      alert('Failed to retrieve recovery phrase. Please try again later.');
+      console.error(e);
+    }
+  };
+
+  const handleCopy = async () => {
+    await copyToClipboard(recoveryPhrase.join(' '));
+    message.success('Recovery phrase copied!');
+    setIsCopied(true);
+    markAsBackedUp();
+
+    setTimeout(() => {
+      setIsCopied(false);
+    }, 2000);
+  };
+
+  const initialDescription =
+    'The Secret Recovery Phrase provides full access to your Pearl wallet and agent funds.';
+  const recoveryPhraseDescription =
+    "Store it in a safe place that only you can access and won't forget.";
+
+  return (
+    <Modal
+      size="small"
+      open={open}
+      onCancel={onClose}
+      title="Secret Recovery Phrase"
+      description={
+        step === 'password' ? initialDescription : recoveryPhraseDescription
+      }
+      closable
+      action={
+        step === 'password' ? (
+          <PasswordStep
+            password={password}
+            setPassword={setPassword}
+            isValidating={isLoading}
+          />
+        ) : (
+          <RecoveryPhraseStep
+            recoveryPhrase={recoveryPhrase}
+            isCopied={isCopied}
+            onCopy={handleCopy}
+          />
+        )
+      }
+      footer={
```

Comment:
I'm a bit confused with that we have the form (PasswordStep) as "action"
I would have expected the actions (buttons) to be in footer, and the content to be in description. but maybe having separate "action" prop in Modal is not needed?